packages/babel-helpers/src/helpers/applyDecs2305.ts

```diff
@@ -674,32 +189,407 @@ function defineMetadata(Class: any, metadata: any) {
+              decInfo[4](this, value);
+            },
+          };
+        } else {
+          desc[key] = decVal;
+        }
+
+        if (!isField) {
+          setFunctionName(
+            desc[key],
+            name,
+            kind === PROP_KIND.METHOD ? "" : key,
+          );
+        }
+      } else if (!isField) {
+        desc = Object.getOwnPropertyDescriptor(Class, name);
+      }
+    }
+
+    var newValue;
+
+    for (var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1) {
+      var dec = (decs as Function[])[i],
+        decThis = decoratorsHaveThis ? (decs as any[])[i - 1] : void 0;
+
+      var decoratorFinishedRef: DecoratorFinishedRef = {};
+      var ctx: DecoratorContext = {
+        kind: [
+          "field",
+          "accessor",
+          "method",
+          "getter",
+          "setter",
+          "field",
+          "class",
+        ][kind] as any,
+
+        name: name,
+        metadata: metadata,
+      };
+
+      ctx.addInitializer = function (
+        decoratorFinishedRef: DecoratorFinishedRef,
+        initializer: Function,
+      ) {
+        if (decoratorFinishedRef.v) {
+          throw new Error(
+            "attempted to call addInitializer after decoration was finished",
+          );
+        }
+        assertCallable(initializer, "An initializer", "be", true);
+        initializers.push(initializer);
+      }.bind(null, decoratorFinishedRef);
+
+      try {
+        if (isClass) {
+          newValue = dec.call(decThis, Class, ctx);
+        } else {
+          ctx.static = isStatic;
+          ctx.private = isPrivate;
+
+          var get, set;
+          if (!isPrivate && (isField || kind === PROP_KIND.METHOD)) {
+            get = function (target: any) {
+              return target[name];
+            };
+            if (isField) {
+              set = function (target: any, v: any) {
+                target[name] = v;
+              };
+            }
+          } else if (kind === PROP_KIND.METHOD) {
+            // Assert: isPrivate is true.
+            get = function (_this: any) {
+              if (isPrivate) {
+                assertInstanceIfPrivate(_this);
+              }
+              return desc.value;
+            };
+          } else {
+            if (kind < PROP_KIND.METHOD || kind === PROP_KIND.GETTER) {
+              get = _bindPropCall(
+                desc,
+                "get",
+                isPrivate && assertInstanceIfPrivate,
+              );
+            }
+            if (kind < PROP_KIND.METHOD || kind === PROP_KIND.SETTER) {
+              set = _bindPropCall(
+                desc,
+                "set",
+                isPrivate && assertInstanceIfPrivate,
+              );
+            }
+          }
+
+          var access: DecoratorContextAccess = (ctx.access = {
+            has: isPrivate
+              ? // @ts-expect-error no thisArg
+                hasPrivateBrand.bind()
```

Comment:
Why do we need to `.bind()`?

Reply:
I don't know, the code just moved. There's even a test to make sure `bind` is called.
![image](https://github.com/babel/babel/assets/30521560/5c300847-897d-4d77-9048-5ca9c5af4003)